feat(grid): add thickness method for all grid types

jdhughes-usgs opened the pull request:
Includes option for passing array to calculate saturated thickness.

jdhughes-usgs commented:
@jlarsen-usgs Added the proposed grid.thick() method for Mike Fienen. 

It also seems the grid.top_botm property was only appropriate for structured grids. 

Let me know what you think of the proposed changes. If it looks good I will edit the pull request and deprecate the thickness property on the ModflowDis class (we should probably do the same for the zcentroids property) and point users to the grid class.

jlarsen-usgs commented on autotest/t076_modelgrid_thick.py:
```diff
@@ -0,0 +1,196 @@
+        [13, 4, 5, 8, 7],
+        [14, 6, 7, 10, 9],
+    ]
+    top = np.ones((nlay, ncpl), dtype=float)
+    top[0, :] = 10.
+    top[1, :] = 5.
+    top[2, :] = 0.
```
```suggestion
    top = np.ones((ncpl, ), dtype=float) * 10
```
Vertex grid follows the Modflow 6 DISV input style of only providing top for the Model top

jlarsen-usgs commented on flopy/discretization/grid.py:
```diff
@@ -275,8 +275,55 @@ def botm(self):
     def top_botm(self):
-        new_top = np.expand_dims(self._top, 0)
-        return np.concatenate((new_top, self._botm), axis=0)
+        elevations = None
+        if self.grid_type == "structured":
+            new_top = np.expand_dims(self._top, 0)
+            elevations = np.concatenate((new_top, self._botm), axis=0)
+        else:
+            raise ValueError(
+                "top_botm property not implemented for vertex "
+                + "and unstructured model grids"
+            )
+        return elevations
```
```suggestion
        new_top = np.expand_dims(self._top, 0)
        return np.concatenate((new_top, self._botm), axis=0)
```

jlarsen-usgs replied:
I'd reccomend keeping this as the original code is Grid(). This works for both DIS and DISV, see DISV top comment above.

I'd reccomend adding 

```
@property
def top_botm(self):
    new_top = np.expand_dims(self._top, 0)
    new_botm = np.expand_dims(self._botm, 0)
    return np.concatenate((new_top, new_botm), axis=0)
```

to the `UnstructuredGrid` class to override base class functionality.

jlarsen-usgs commented on flopy/discretization/grid.py:
```diff
@@ -275,8 +275,55 @@ def botm(self):
+        -------
+            thick : calculated thickness
+        """
+        if self.grid_type == "structured":
+            thick = -np.diff(self.top_botm, axis=0)
+            if array is not None:
+                top = self.top_botm[:-1]
+                bot = self.top_botm[1:]
+                idx = (array < top) & (array > bot)
+                thick[idx] = array[idx] - bot[idx]
+                idx = array <= bot
+                thick[idx] = 0.0
+        else:
+            thick = self.top - self.botm
+            if array is not None:
+                idx = (array < self.top) & (array > self.botm)
+                thick[idx] = array[idx] - self.botm[idx]
+                idx = array <= self.botm
+                thick[idx] = 0.0
+        return thick
```
```suggestion
        thick = -np.diff(self.top_botm, axis=0)
        if array is not None:
            array.shape = thick.shape
            top = self.top_botm[:-1]
            bot = self.top_botm[1:]
            idx = np.where((array < top) & (array > bot))
            thick[idx] = array[idx] - bot[idx]
            idx = np.where(array <= bot)
            thick[idx] = 0.0
        return thick
```

Once top_botm property is added to the UnstructuredGrid class this can be simplified as so

jlarsen-usgs reviewed (commented):
Hi Joe,

I've added some general comments and changes that will help remove the if/else statements from the code. The main being implementing a top_botm property method in the UnstructuredGrid class that overrides the base Grid class.

jdhughes-usgs commented:
@jlarsen-usgs I implemented your suggestions. I modified them a bit because the unstructured was returning an array for the thickness instead of a vector. So thick return arrays with (nlay, nrow. ncol), (nlay, ncpl), and (nodes) shapes for structured, vertex, and unstructured grids.

Also made thick a property and added a separate saturated_thick method for calculating the saturated thickness.

jdhughes-usgs commented:
@aleaf I was working on a generic thick property for the grid class based on a conversation I had with @mnfienen yesterday. I also added a generic saturated_thick method. 

I was going to replace the use of get_saturated_thickness in t042 with the new grid class method. I noticed that the current get_saturated_thickness method calculates a value of 1.6 for cell (2,1,1) which is has top and bottom elevations of 2 and 1, respectively, and a head of 2.6. This isn't what I would call the saturated thickness which to me is 1.0. Instead I would say what get_saturated_thickness is calculating is a pressure head. 

I was hoping to deprecate the old dis.thickness property and postprocessing.get_saturated_thickness method in 3.3.5 but couldn't do it for get_saturated_thickness since it is not equivalent to what grid.saturated_thick is calculating. Do you have any thoughts.

jdhughes-usgs commented:
@jlarsen-usgs You should also look at the t050 and t072 autotests (I think you are the one who worked on the parts of the autotests that I modified). I had to change some of the values in the assertions. The specific discharge values are different because I replace the postprocessing,get_saturated_thickness() method with the grid.saturated_thick() method. For cases where the head is above the top of the cell the postprocessing,get_saturated_thickness() method is calculating the pressure head not the aquifer thickness. So the fact that the vectors are changes does not surprise me but you should look at it to confirm this.

jlarsen-usgs commented:
@jdhughes-usgs 

I looked through the new changes for calculating saturated thickness and the associated tests and they make sense to me. The changes in vector values are what I would expect from changing from potentiometric surface to layer thickness when heads are higher than a given model cell's top elevation.
